doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,350 @@
+
+### More Keys
+
+Even if calling `kdbGet` with a parent key below a mountpoint, `kdbGet` will nevertheless return all keys of the mountpoint.
+Pseudo code example, assuming there is a mountpoint at `/mountpoint` and a key `/mountpoint/other`:
+
+```
+kdbGet (kdb, ks, keyNew("/mountpoint/below"));
+assert (ksLookup (ks, "/mountpoint/other") == NULL);
+```
+
+It was found unexpected that this assert will fail.
+
+### Fewer Keys
+
+When doing a second `kdbGet` with a new keyset no keys will be returned when no backends report changed data, because kdb internally thinks the data is already up-to-date.
+A unit test by @atmaxinger:
+
+```c
+static void test_doubleGet (void)
+{
+	printf("running %s\n", __func__);
+
+	// Setup
+	Key * parentKey = keyNew("/somewhere", KS_END);
+	KeySet * ks = ksNew(0, KS_END);
+	KDB * kdb = kdbOpen (ksNew(0, KS_END), parentKey);
+	kdbGet (kdb, ks, parentKey);
+	ksAppendKey (ks, keyNew ("user:/somewhere", KEY_VALUE, "abc", KEY_END));
+	ksAppendKey (ks, keyNew ("user:/somewhere/key", KEY_VALUE, "xyz", KEY_END));
+	kdbSet (kdb, ks, parentKey);
+	kdbClose (kdb, parentKey);
+
+	// Scenario
+	kdb = kdbOpen (ksNew(0, KS_END), parentKey);
+
+	KeySet * ks1 = ksNew (0, KS_END);
+	KeySet * ks2 = ksNew (0, KS_END);
+
+	kdbGet (kdb, ks1, keyNew("/somewhere", KEY_END));
+	succeed_if (ksLookupByName (ks1, "/somewhere/key", 0) != NULL, "should find key (1)");
+	kdbGet (kdb, ks2, keyNew("/somewhere", KEY_END));
+	succeed_if (ksLookupByName (ks2, "/somewhere/key", 0) != NULL, "should find key (2)");
+}
+```
+
+> It actually outputs should find key (2) so the assertion fails.
+
+It was found unexpected that the second assert will fail.
+
+## Constraints
+
+- memory consumption must be low for `kdbGet`, see [4. Goal: Performance](/doc/GOALS.md), in particular, deep duplication is too expensive
+
+## Assumptions
+
+## Considered Alternatives
+
+### Keep Current Situation
+
+Improve documentation to make people more aware of these two problems:
+
+- add a tutorial about `kdbGet` semantics
+- add full examples how to correctly work with `kdbGet`
+
+### Cachefilter Plugin
+
+Naively one would simply cache the whole keyset and use `ksBelow` to always get the keyset.
+
+This idea was implemented and later on discarded: a3d95f07160d792fdd0f169d8543138c32a2f580
+
+The main problems are:
+
+- very high memory consumption (duplication of KeySets)
+- problems specific to [hooks](../4_partially_implemented/hooks.md), see [#1072](https://issues.libelektra.org/1072)
+
+### MMAP Cache with parent key
+
+We make the mmap cache non-optional so that we always have a keyset of configuration data internally.
+From this keyset, we use `ksBelow` to return the correct keyset.
+
+**Cons:**
+
+- invalidation of OPMPHM
+
+### MMAP Cache without parent key
+
+We make the mmap cache non-optional and only use a single cache, caching everything.
+We remove the parent key of `kdbGet` and `kdbSet` and always return the keyset of the whole KDB.
+
```

Comment:
```suggestion
Advantage: then also symlinks and constraints for keys outside of the parentKey would work.

```

Reply:
Would that really be a good thing? Reach up out of the mountpoint breaks isolation in any case. It can also only work, within hook plugins, because mountpoints are still isolated for mounted plugins.

Reply:
symlinks are not affected by the problem you describe. I updated the suggestions to make this more clear. I agree that this is actually a separate problem but it would be solved by some mmap solutions proposed here.

Reply:
What exactly do you mean by "symlinks"?

`kdb get -a` is exactly the issue I describe. I normally expect mountpoints to be isolated. Having a key `system:/foo/bar` refer to `../../abc` (in whatever way) when `system:/foo` is a mountpoint is always fishy. The other way around e.g. `system:/abc` referring to `system:/foo/bar` (assuming `system:/abc` is _not_ a mountpoint) would be less fishy, but still a bit weird.

Reply:
Is there a resolution to this?

Reply:
I updated the suggestion already before, it explains symlinks now.

The other stuff is imho another decision and not really important as symlinks already show the whole problem.

Reply:
You still haven't addressed why this would be actually be a good default behaviour. You just claim it is an advantage without any explanation.

Like I said, I would normally expect mountpoints to be isolated. If symlinks work like this, the isolation is partially broken. One could argue that the problem right now is that such "broken" symlinks are not prevented. But your solution doesn't completely fix the problem either.

Take a variant of my previous example:
- `system:/foo` is a mountpoint, there are no other mountpoints
- `spec:/foo/bar` refers to `system:/abc` via `meta:/override`

Now in a plugin that is part of the mountpoint `system:/foo` doing a `ksLookupByName (ks, "/foo/bar", 0)` would **NOT** use `system:/abc`, because that key is **never** part of the keyset passed to this plugin.

However, in an application that used `system:/foo` as the parent, the `meta:/override` **would work** with your proposal. To me that seems like very confusing behaviour, because both the application and the plugin seemingly use the same parent key.

Reply:
This is a separate decision. Can you please create one?

Reply:
I added this discussion to the decision.